service/awsconfig/v10/resource/cloudformation/update.go

```diff
@@ -22,7 +27,37 @@ func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange inte
 	if updateStackInput.StackName != nil && *updateStackInput.StackName != "" {
 		r.logger.LogCtx(ctx, "level", "debug", "message", "updating the guest cluster main stack")
 
-		_, err := r.clients.CloudFormation.UpdateStack(&updateStackInput)
+		// Fetch the etcd volume information.
+		etcdVolume := true
+		persistentVolume := false
+		volumes, err := r.ebs.ListVolumes(customObject, etcdVolume, persistentVolume)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(volumes) != 1 {
+			return microerror.Maskf(executionFailedError, "there must be 1 volume for etcd, got %d", len(volumes))
+		}
+		vol := volumes[0]
+
+		// First shutdown the instances and wait for it to be stopped. Then detach
+		// the etcd volume without forcing.
+		force := false
+		shutdown := true
+		wait := true
+		for _, a := range vol.Attachments {
+			err := r.ebs.DetachVolume(ctx, vol.VolumeID, a, force, shutdown, wait)
+			if err != nil {
+				r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("failed to detach EBS volume %s", vol.VolumeID), "stack", fmt.Sprintf("%#v", err))
+			}
+		}
+
+		// Once the etcd volume is cleaned up and the master instance is down we can
+		// go ahead to let CloudFormation do its job.
+		//
+		// NOTE the update proceeds even if the volume detachements above fail. We
+		// keep going to update what we are able to even if master nodes are not
+		// updated in error cases.
```

Comment:
This seems a bit scary. What happens if the detachment fails but the AMI has changed? The master will be replaced but what if it won't attach to the new instance?

Would it be safer to stop and return an execution failed error? So that we can then investigate manually.

Reply:
Makes sense. More defensive now. We return the error.